### PR TITLE
[Model] Optimize MiniCPM-o streaming audio encoder memory

### PR DESCRIPTION
**Pull request by xsmccc**
Related to #5069.

This is the focused streaming-audio follow-up discussed in
https://github.com/vllm-project/vllm-omni/issues/5069#issuecomment-5119179172
and https://github.com/vllm-project/vllm-omni/pull/5188#issuecomment-5033947603.

### Problem

`get_audio_embedding_streaming()` always requested
`output_hidden_states=True` and selected `hidden_states[-1]`, even when
`audio_encoder_layer == -1` means that only Whisper's final encoder output is
needed. This retains the embedding output and every intermediate encoder-layer
state on every cached streaming chunk.

### Changes

- Use `last_hidden_state` and set `output_hidden_states=False` for the default
  final-layer selection (`audio_encoder_layer == -1`).
- Preserve the existing `hidden_states[audio_encoder_layer]` behavior for
  non-final layer selection, including Whisper's embedding-output-first hidden
  state ordering.
- Leave the streaming KV-cache, attention mask, extra-context arguments,
  projection, and pooling paths unchanged.
- Add focused CPU regression tests for:
  - final-layer selection without retaining intermediate hidden states;
  - non-final layer selection with hidden states enabled;
  - two cached chunks with cache reuse, growing attention-mask/cache lengths,
    and unchanged extra-context arguments.

The production diff is intentionally small. The one-off GPU comparison harness
and result artifacts are not included in this PR, following the review guidance
from #5188.

## Test Plan

**vLLM Version:** 0.25.0

**vLLM-Omni Commit:** `ec760d9bd8448c1c7a8b34778c62408880496076`

**Base Commit:** `1d6c0816a4fbd9079878d62bbe28c1a12f146a1f`

The GPU evidence was captured before the final rebase. Git object verification
shows that the MiniCPM-o production file is byte-identical across the old and
current bases (blob `577ce5aa`) and across the benchmarked and current
candidates (blob `d37aa0d1`).

**Environment:**

- Python 3.12.3
- PyTorch 2.11.0+cu130
- Transformers 5.14.1
- CUDA runtime reported by PyTorch: 13.0
- NVIDIA driver 590.44.01
- NVIDIA RTX PRO 6000 Blackwell Server Edition, 96 GiB
- Model: `OpenBMB/MiniCPM-o-4_5`, BF16, eager attention

Commands:

```bash
pytest -q tests/model_executor/models/minicpmo_4_5/test_audio_streaming.py
pytest -q \
  tests/model_executor/models/minicpmo_4_5 \
  tests/worker/test_native_duplex_hooks.py

pre-commit run debug-statements --from-ref origin/main --to-ref HEAD
pre-commit run end-of-file-fixer --from-ref origin/main --to-ref HEAD
pre-commit run mixed-line-ending --from-ref origin/main --to-ref HEAD
pre-commit run trailing-whitespace --from-ref origin/main --to-ref HEAD
pre-commit run ruff-check --from-ref origin/main --to-ref HEAD
pre-commit run ruff-format --from-ref origin/main --to-ref HEAD
pre-commit run typos --from-ref origin/main --to-ref HEAD
pre-commit run check-test-ci-coverage --from-ref origin/main --to-ref HEAD
pre-commit run check-pickle-imports --from-ref origin/main --to-ref HEAD

PATH=/usr/bin:$PATH bash tools/run_ready_jobs.sh --dry-run
```

The GPU comparison used a fixed 8.03-second, 16 kHz audio prefix split into
eight approximately one-second cached chunks. Each process ran one unmeasured
warm-up streaming session, one peak-memory session, and then latency
repetitions. The complete baseline/candidate pair was run twice independently.

### CPU and repository checks

- Focused streaming tests: `3 passed, 14 warnings in 0.40s`
- MiniCPM-o model-executor tests plus native duplex hooks:
  `173 passed, 1 skipped, 17 warnings in 2.68s`
- All listed pre-commit hooks passed.
- `check-test-ci-coverage` passed, and `tools/run_ready_jobs.sh --dry-run`
  places the new tests in the `Simple Model Executor Test` CPU job.

### Cached streaming numerical equivalence

Both independent baseline/candidate pairs produced:

- 327,680 output elements bit-exact (`max_abs_error = 0`,
  `mean_abs_error = 0`, cosine similarity `1.0`);
- all 48 final KV-cache tensors bit-exact;
- matching cache lengths of 50, 100, ..., 400 tokens;
- reuse of the same cache object after the first chunk.

### Peak GPU memory

| Run | Baseline peak allocated delta | PR peak allocated delta | Saving | Baseline peak reserved delta | PR peak reserved delta | Saving |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | 41.72 MiB | 40.76 MiB | 0.96 MiB | 54 MiB | 50 MiB | 4 MiB |
| 2 | 41.72 MiB | 40.76 MiB | 0.96 MiB | 56 MiB | 50 MiB | 6 MiB |

As a non-regression sanity check, mean eight-chunk latency was
63.38 ms vs. 62.63 ms over 20 repetitions in run 1, and
66.11 ms vs. 62.63 ms over 5 repetitions in run 2. These sequential local
measurements are not presented as a latency improvement claim.

### Local environment limitation

Transformers 5.14.1 has cache-interface differences from the copied Whisper
cache contract used by this checkout. The out-of-tree GPU harness therefore
applied identical compatibility adapters to baseline and candidate for
`DynamicCache.get_usable_length`, `EncoderDecoderCache` indexing, and the
Whisper attention cache call/return contract. The relative numerical and memory
comparison remains controlled; no compatibility adapter is part of this PR.

The broader local command
`pytest -q tests/model_executor -m "core_model and cpu"` is also blocked during
collection in unchanged Qwen model tests because the installed vLLM 0.25.0 does
not export `vllm.multimodal.utils.set_mm_embedding_modality`. The focused and
related MiniCPM-o suites above complete successfully; project CI should run the
full split in its pinned environment.

### CODE REVIEW

**Comment by amy-why-3459** [comment]
Docs build is failing, please run cd docs && make html to locate and fix the error.

**Comment by xsmccc** [comment]
> 文档构建失败了，请运行 CD 文档并制作 HTML 来定位并修复错误。

Thanks for flagging this. I checked the exact PR head, `8c417ddfca490b60bb608a24c59fe92bfd0b58ec`.

The Read the Docs build did not reach the MkDocs build step. Its detailed notification reports:

`Error while checking out the repository`  
`Failed to checkout revision: 8c417ddfca490b60bb608a24c59fe92bfd0b58ec`

On the current branch, documentation is configured through the root `.readthedocs.yml` and `mkdocs.yml`; there is no `docs/Makefile`, so I ran the equivalent strict build on the exact PR head:

`python -m mkdocs build --strict`

It completed successfully: `Documentation built in 205.83 seconds`.

The PR changes only the MiniCPM-o model implementation and its test file, with no documentation or documentation-configuration changes. Since the PR head revision is now available, this appears to be a transient checkout/ref-propagation failure after syncing `main`. Could someone with Read the Docs access please rerun that build?
